### PR TITLE
feat: add k8s adapter

### LINKED ISSUE #1633: [Feature] Add K8s adapter to ConfigLoader
### The problem faced currently?

We don't have an ConfigLoader adapter for Kubernetes

### How can we solve it?

Make one.

Notes:
1. Use single shared informer for watching all GVKs.
2. We will need a list of GVKs while initializing this adapter.

>If you want this feature to be implemented, give it a thumbs up reaction, so that we can determine which features are important to you.
>👍

### PR DESCRIPTION
**Pull request by shubham4443**
Fixes #1633

### CODE REVIEW

File: sc-poc/managers/configman/adapter/k8s/k8s.go [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	dc, err := dynamic.NewForConfig(restConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	factory := dynamicinformer.NewFilteredDynamicSharedInformerFactory(dc, 5*time.Minute, v1.NamespaceAll, nil)
```
Comment (YourTechBud) [inline sc-poc/managers/configman/adapter/k8s/k8s.go:54]:
```suggestion
	factory := dynamicinformer.NewFilteredDynamicSharedInformerFactory(dc, 5*time.Minute, namespace, nil)
```

File: sc-poc/managers/configman/adapter/k8s/k8s.go [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+				for i, spec := range k.configurationN[key] {
+					if spec.GetName() == u.GetName() {
+						k.configurationN[key][i] = u
+						found = true
+						break
+					}
+				}
+				if !found {
+					k.configurationN[key] = append(k.configurationN[key], u)
+				}
```
Comment (YourTechBud) [inline sc-poc/managers/configman/adapter/k8s/k8s.go:106]: Make this a helper function `addOrUpdateConfig`

File: sc-poc/managers/configman/adapter/k8s/k8s.go [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+				if !found {
+					k.configuration[kind] = append(k.configuration[kind], u)
+				}
+
+				resp, err := k.toRaw()
+				if err != nil {
+					k.logger.Error("error reloading config: ", zap.Error(err))
+					return
+				}
```
Comment (YourTechBud) [inline sc-poc/managers/configman/adapter/k8s/k8s.go:126]: Put in function `addOrUpdateOldConfig`

File: sc-poc/managers/configman/adapter/k8s/k8s.go [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+				// new configuration
+				key := source.GetModuleName(u.GetAPIVersion(), u.GetKind())
+				s := []*unstructured.Unstructured{}
+				for _, spec := range k.configurationN[key] {
+					if spec.GetName() == u.GetName() {
+						continue
+					}
+					s = append(s, spec)
+				}
+				k.configurationN[key] = s
```
Comment (YourTechBud) [inline sc-poc/managers/configman/adapter/k8s/k8s.go:171]: I suggest you use this method to remove to minimize allocs 
```golang
func remove(s []int, i int) []int {
    s[i] = s[len(s)-1]
    return s[:len(s)-1]
}
```

File: sc-poc/managers/configman/adapter/k8s/k8s.go [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+			},
+		})
+
+		go informer.Run(ctx.Done())
+	}
+
+	return cfgChan, nil
+}
+
+func (k *K8s) toRaw() ([]byte, error) {
```
Comment (YourTechBud) [inline sc-poc/managers/configman/adapter/k8s/k8s.go:200]: Rename to `getCaddyConfig()`

**Review by YourTechBud — CHANGES_REQUESTED** [review]

File: sc-poc/managers/configman/adapter/k8s/k8s.go (hunk 4 above)
Comment (shubham4443) [reply sc-poc/managers/configman/adapter/k8s/k8s.go:128]: this method doesn't work when length is 1

**Review by YourTechBud — APPROVED** [review]

**Pull request merged**